### PR TITLE
Refactoring gmsh issue #15

### PR DESCRIPTION
**Pull request by ezrayst**

### CODE REVIEW

File: include/gmsh.tcc [hunk 1]
```diff
@@ -4,137 +4,149 @@
...
+//! \param[in] keyword search for the word of interest
+template <unsigned Tdim, unsigned Tvertices>
+void GMSH<Tdim, Tvertices>::read_keyword(std::ifstream& file,
+                                         const std::string& keyword) {
+
+  bool read_status = false;
+  std::string line;
+  file.clear();
+  file.seekg(0, std::ios::beg);
+  while (std::getline(file, line)) {
```
Comment (kks32) [inline include/gmsh.tcc]: Needs more documentation on what this is doing.

File: include/gmsh.tcc [hunk 2]
```diff
@@ -4,137 +4,149 @@
...
-    getline(infile, line);
-    //! Skip past vertices section
-    for (int i = 0; i < nvertices; ++i) {
-      std::getline(infile, line);
-    }
-    //! Skip past element headers
-    for (unsigned l = 0; l < 2; ++l) {
-      std::getline(infile, line);
-    }
+  //! specify element type 4 = quadrilateral, 8 = hexahedron
```
Comment (kks32) [inline include/gmsh.tcc]: Add reference to GMSH doc showing element type

File: include/gmsh.tcc [hunk 3]
```diff
@@ -227,19 +239,26 @@ void GMSH<Tdim, Tvertices>::compute_material_points() {
...
-  const double max_height = 3;
   const double conv_factor = 10;
 
-  std::array<double, Tdim> stresses;
+  double max_height = std::numeric_limits<double>::min();
+
+  //! [2D], y is the vertical direction
+  //! [3D], z is the vertical direction
+  //! In general, [Tdim - 1]
+  for (auto const& point : materialpoints_) {
```
Comment (kks32) [inline include/gmsh.tcc]: `const auto`

File: include/gmsh.tcc [hunk 4]
```diff
@@ -227,19 +239,26 @@ void GMSH<Tdim, Tvertices>::compute_material_points() {
 template <unsigned Tdim, unsigned Tvertices>
 void GMSH<Tdim, Tvertices>::compute_stresses() {
```
Comment (kks32) [inline include/gmsh.tcc:260]: Should compute_stress be part of mesh?

**Review by kks32 — CHANGES_REQUESTED** [review]

File: include/gmsh.tcc (hunk 4 above)
Comment (ezrayst) [reply include/gmsh.tcc:260]: I don't think so... so where to put? I personally recommend to move it to another header file if possible

File: include/gmsh.tcc [hunk 5]
```diff
@@ -4,137 +4,166 @@
...
+  //! 3 - Quadrangle (4 nodes)
+  //! 4 - Tetrahedron (4 nodes)
+  //! 5 - Hexahedron (8 nodes)
+  //! 6 - Prism (6 nodes)
+  //! 7 - Pyramid (5 nodes)
+  //! 8 - Second order line (3 nodes)
+  //! 9 - Second order triangle (6 nodes)
+  //! 11 - Second order tetrahedron (10 nodes)
+  //! 15 - Point (1 node)
+  const unsigned element_type = 4;
```
Comment (kks32) [inline include/gmsh.tcc]: This conflicts with the update related to Hexahedron.

Comment (kks32) [reply include/gmsh.tcc]: Rebase to develop to fix this.

**Review by kks32 — CHANGES_REQUESTED** [review]

**Review by kks32 — APPROVED** [review]

**Pull request merged**